docs/topics/functions.md

```diff
@@ -135,73 +240,90 @@ You can skip all the ones with default values:
-_spread_ operator (prefix the array with `*`):
+You can pass a [variable number of arguments](#variable-number-of-arguments-varargs) (`vararg`) naming the correspoding array:
 
 ```kotlin
-fun foo(vararg strings: String) { /*...*/ }
+fun mergeStrings(vararg strings: String) { /*...*/ }
 
-foo(strings = *arrayOf("a", "b", "c"))
+mergeStrings(strings = arrayOf("a", "b", "c"))
 ```
 
+<!-- Rationale for named arguments interaction with varargs is here https://youtrack.jetbrains.com/issue/KT-52505#focus=Comments-27-6147916.0-0 -->
+
 > When calling Java functions on the JVM, you can't use the named argument syntax because Java bytecode does not
 > always preserve the names of function parameters.
 >
 {style="note"}
 
-### Unit-returning functions
-
-If a function does not return a useful value, its return type is `Unit`. `Unit` is a type with only one value - `Unit`.
-This value does not have to be returned explicitly:
-
-```kotlin
-fun printHello(name: String?): Unit {
-    if (name != null)
-        println("Hello $name")
-    else
-        println("Hi there!")
-    // `return Unit` or `return` is optional
-}
-```
+### Explicit return types
 
-The `Unit` return type declaration is also optional. The above code is equivalent to:
+Functions with a block body must always specify return types explicitly, unless it's intended for them to return `Unit`,
+[in which case specifying the return type is optional](#unit-returning-functions).
 
-```kotlin
-fun printHello(name: String?) { ... }
-```
+Kotlin does not infer return types for functions with block bodies because such functions may have complex control flow
+in the body, and the return type will not be obvious to the reader (and sometimes even for the compiler).
+But the return type is inferred, unless specified, for [single-expression functions](#single-expression-functions).
 
 ### Single-expression functions
 
-When the function body consists of a single expression, the curly braces can be omitted and the body specified after an `=` symbol:
+When the function body consists of a single expression, you can omit the curly braces and specify the body after an `=` symbol:
 
 ```kotlin
 fun double(x: Int): Int = x * 2
 ```
 
-Explicitly declaring the return type is [optional](#explicit-return-types) when this can be inferred by the compiler:
+You don't have to [explicitly declare](#explicit-return-types) the return type when it can be inferred by the compiler:
 
 ```kotlin
 fun double(x: Int) = x * 2
 ```
 
-### Explicit return types
+### Unit-returning functions
 
-Functions with block body must always specify return types explicitly, unless it's intended for them to return `Unit`,
-[in which case specifying the return type is optional](#unit-returning-functions).
+If a function has a block body and does not return a useful value, compiler assumes the return type to be `Unit`
+(corresponds to the `void` type in Java).
```

Comment:
Can you remove this part about Java? I'd like to avoid specifically comparing with Java since we're trying to keep our docs as language-agnostic as possible :)